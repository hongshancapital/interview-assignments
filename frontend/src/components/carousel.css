.caption {
	top:  20%;
}

.caption h1 {
	font-size: 4.5em;
}

.caption p {
	font-size: 2em;
}

.black-font {
	color: black !important;
}

.iphone {
	height:100vh;

	background-image: url("../assets/iphone-16-9.png");
	background-repeat: no-repeat !important;
  background-size: 100% 100% !important;
}

.tablet {
	height:100vh;

	background-image: url("../assets/tablet-16-9.png");
	background-repeat: no-repeat !important;
  background-size: 100% 100% !important;
}

.airpods {
	height:100vh;

	background-image: url("../assets/airpods-16-9.png");
	background-repeat: no-repeat !important;
  background-size: 100% 100% !important;
}

.progress-bars {
	position: absolute;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 15;
  display: flex;
  justify-content: center;
  padding-left: 0;
  margin-right: 15%;
  margin-left: 15%;
  list-style: none;
}

.progress-bars li {
	width: 30px;
	margin-left: 3px;
	margin-right: 3px;
}

.progress {
	height: 3px !important;
	background-color: #bfbfbf !important;
	pointer-events: auto;
  cursor: pointer;
}

.progress-bar {
	background-color: #8c8c8c !important;
}
