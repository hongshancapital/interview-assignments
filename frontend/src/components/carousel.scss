/* div{
    color: red;
} */
.box {
    width: 200px;
    height: 200px;
    margin: auto 0;
    border: 1px solid blue;
    overflow: hidden;
}

.box-item {
    width: 600px;
    height: 200px;
    transform: translateX(0);
    transition: transform 1s;
    overflow: hidden;
}
.ca-item {
    width: 200px;
    height: 200px;
    color: white;
    display: inline-block;
    line-height: 200px;
    text-align: center;
}
.ca-item-0, .ca-item-3 {
    background-color: rgb(167, 104, 104);
}
.ca-item-1 {
    background-color: rgb(42, 44, 190);
    display: inline-block;
}
.ca-item-2 {
    background-color: rgb(143, 187, 41);
    display: inline-block;
}