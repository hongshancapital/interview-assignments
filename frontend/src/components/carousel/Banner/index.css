.banner_control{
  width: 100%;
  height: 100%;
  padding-top:210px;
  box-sizing: border-box;
  background-size: auto 133px;
  background-repeat: no-repeat;
  background-position: center 71.6%;
  position: absolute;
}

.banner_control .title{
  margin-left: auto;
  margin-right: auto;
}

.banner_control .text{
  margin-left: auto;
  margin-right: auto;
}

.banner_control.dark .title,
.banner_control.dark .text{
  color:#fff;
}