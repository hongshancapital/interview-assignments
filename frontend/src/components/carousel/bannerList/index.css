.banner-list-control {
  width: 100%;
  height: 100%;
  position: absolute;
  left: 0;
  top: 0;
  transition: left 0.5s;
}
