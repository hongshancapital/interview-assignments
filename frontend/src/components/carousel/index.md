旋转木马，一组轮播的区域。

## API
| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- | --- |
| autoplay | 是否自动切换 | boolean | true |
| afterChange | 切换到指定面板 | function(index) | - |
| className |类名——用来自定义样式|string|-|


## 实现思路
拆分为两部分实现：轮播动画播放部分和轮播按钮部分。
轮播动画主要移动父元素，从而实现子元素轮播的效果。
动画通过slider值变化和`translateX(-${(100 * slider) / childLength}%)`实现。
自动播放时，3s后slider+1,监听到slider变化，自动执行后面的动画；手动播放先清除定时器，然后slider赋值为当前点击的slider。
可通过传入className随意修改组件样式。