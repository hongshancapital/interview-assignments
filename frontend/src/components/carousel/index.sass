
.carousel-main
  display: flex
  flex-direction: row
  overflow: hidden

  .carousel-item
    height: 100vh
    width: 100vw
    overflow: hidden
    

.carousel-dots-group
  display: flex
  position: fixed
  top: 91%
  left: 50%
  transform: translateX(-50%)

  .carousel-dots-item
    width: 6.2rem
    margin: 0 1rem
    height: 4px
    border-radius: 4px
    background: #ccc
    cursor: pointer

    .carousel-dots-item-dot
      background: #fff
      border-radius: 4px
      height: 4px
      width: 0%

      &.active
        @keyframes dot-active 
            0%   
              width: 0%
            100% 
              width: 100%
