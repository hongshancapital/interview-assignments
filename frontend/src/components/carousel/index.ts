import Carousel from "./Carousel";

export default Carousel;