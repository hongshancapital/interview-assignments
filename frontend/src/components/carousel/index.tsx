import React from 'react';
import { SlideItem } from '../slide-item';
import { useCarousel } from './hook';
import './index.css';

export const Carousel: React.FC = () => {
  const { curIndex, slideItems, animationConfig, changeIndicator } = useCarousel();
  const { duration, transitionDuration } = animationConfig;

  /** 渲染滑块区 */
  const renderContent = () => {
    return (
      <div
        className='swiper-wrapper'
        style={{
          transitionDuration: transitionDuration + 'ms',
          transform: `translateX(${-curIndex * 100}%)`
        }}
        data-testid='swiper-wrapper'
      >
        {slideItems.map((item) => {
          const { slideId, bgColor, imgUrl, descContents } = item;
          return (
            <div key={slideId} className='swiper-item' style={{ backgroundColor: bgColor }}>
              <SlideItem imgUrl={imgUrl} descContents={descContents}/>
            </div>
          )
        })}
      </div>
    );
  }

  /** 渲染锚点指示器 */
  const renderIndicator = () => {
    return (
      <div className='indicator-wrapper' data-testid='indicator-wrapper'>
        {slideItems.map((item, index) => {
          return (<div
            key={item.slideId}
            className='indicator-item'
            onClick={() => changeIndicator(index)}
          >
              <div
                className='indicator-process'
                style={
                  index === curIndex
                    ? { width: '50px', transition: `${duration}ms ease-in-out` } 
                    : {}
              } />
          </div>);
        })}
      </div>
    );
  }

  return <div className='carousel'>
    {renderContent()}
    {renderIndicator()}
  </div>;
}
