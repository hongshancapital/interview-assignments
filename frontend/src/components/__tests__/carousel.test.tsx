import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Carousel } from '../carousel';

describe('test carousel', () => {
  it('carousel snapshot', () => {
    // 渲染
    const { container } = render(<Carousel />);
    // 快照记录
    expect(container).toMatchSnapshot();
  });

  it('carousel elements', () => {
    const { getByTestId } = render(<Carousel />);
    const swiperWrapper = getByTestId('swiper-wrapper');
    const indicatorWrapper = getByTestId('indicator-wrapper');
    expect(swiperWrapper).toBeInTheDocument();
    expect(indicatorWrapper).toBeInTheDocument();
  });

  it('carousel indicator click', () => {
    const { getByTestId } = render(<Carousel />);
    const indicatorWrapper = getByTestId('indicator-wrapper');
    const indicatorElementTwo = indicatorWrapper.lastElementChild;
    fireEvent.click(indicatorElementTwo!);
    expect(indicatorElementTwo!.firstChild).toHaveStyle({
      transition: `3000ms ease-in-out`,
    });
  });
});