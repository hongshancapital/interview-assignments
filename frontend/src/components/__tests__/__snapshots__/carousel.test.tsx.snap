// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test carousel carousel snapshot 1`] = `
<div>
  <div
    class="carousel"
  >
    <div
      class="swiper-wrapper"
      data-testid="swiper-wrapper"
      style="transition-duration: 1000ms; transform: translateX(0%);"
    >
      <div
        class="swiper-item"
      >
        <div
          class="slide-item"
        >
          <img
            alt="slide-img"
            class="slide-img"
            src="iphone.png"
          />
          <div
            class="slide-desc"
            data-testid="slide-desc"
          >
            <div
              style="color: white; font-size: 40px; font-weight: bold; line-height: 60px;"
            >
              xPhone
            </div>
            <div
              style="color: white; font-size: 30px; font-weight: normal; line-height: 40px; margin-top: 15px;"
            >
              Lots to love.Less to spend.
            </div>
            <div
              style="color: white; font-size: 30px; font-weight: normal; line-height: 40px;"
            >
              Starting at $399.
            </div>
          </div>
        </div>
      </div>
      <div
        class="swiper-item"
      >
        <div
          class="slide-item"
        >
          <img
            alt="slide-img"
            class="slide-img"
            src="tablet.png"
          />
          <div
            class="slide-desc"
            data-testid="slide-desc"
          >
            <div
              style="color: black; font-size: 40px; font-weight: bold; line-height: 60px;"
            >
              Tablet
            </div>
            <div
              style="color: black; font-size: 30px; font-weight: normal; line-height: 40px; margin-top: 15px;"
            >
              Just the right amount of everything.
            </div>
          </div>
        </div>
      </div>
      <div
        class="swiper-item"
      >
        <div
          class="slide-item"
        >
          <img
            alt="slide-img"
            class="slide-img"
            src="airpods.png"
          />
          <div
            class="slide-desc"
            data-testid="slide-desc"
          >
            <div
              style="color: black; font-size: 40px; font-weight: bold; line-height: 60px;"
            >
              Buy a Tablet or xPhone for college.
            </div>
            <div
              style="color: black; font-size: 40px; font-weight: bold; line-height: 60px;"
            >
              Get airPods.
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="indicator-wrapper"
      data-testid="indicator-wrapper"
    >
      <div
        class="indicator-item"
      >
        <div
          class="indicator-process"
          style="width: 50px; transition: 3000ms ease-in-out;"
        />
      </div>
      <div
        class="indicator-item"
      >
        <div
          class="indicator-process"
        />
      </div>
      <div
        class="indicator-item"
      >
        <div
          class="indicator-process"
        />
      </div>
    </div>
  </div>
</div>
`;
